### ISSUE TITLE
asyncshell (via mpd widget) leaves (a lot of) files behind in /tmp/asyncshell

### ISSUE THREAD

**Issue by Unode** [copycat-killer/lain#120]
I noticed this after enabling the mpd widget that uses asyncshell.

In a machine that is on almost 24/7 there were over 100k files in `/tmp/asyncshell` totalling ~40Mb. `/tmp/` is configured to use tmpfs.
One file is created every time the mpd widget updates which is about once per second.

These files are named `reqXXXXX` where `XXXXX` seems to be some kind of incremental identifier.

**Comment by copycat-killer** [comment]
#98

**Comment by Unode** [comment]
I'd close this given it's a duplicate but the answer in the other issue is kinda not really an answer.
Is there really no other way of doing this? Perhaps even a way to cleanup the files afterwards?
Given they seem sequential, couldn't the next job cleanup the files from a few runs (say 50) back if they exist?

I have no problem with temporary files, but not when they accumulate faster than tmpreaper or similar programs can clean them.

**Comment by copycat-killer** [comment]
Not really an answer? That's how it works, I wonder what else do you expect me to tell.

And yes, that's the current drawback for being asynchronous.

One way would be to make `asyncshell` to use a limited number of PIDs, like Linux core does, or to overwrite the same ID when a command is reiterated.

But that requires time, because I'm very busy in this period.

**Comment by Unode** [comment]
Sorry if I wasn't clear. First of all thanks for the time you took to reply. We alI know what busy means and this is free-time/hobby so Thanks!.

What I meant by not really an answer was that the reply was a little dismissive of the problem. And I think a reasonable compromise could be made: https://gist.github.com/Unode/71b9d9a60b474ba1679f
This simply gets rid of the files and the table reference once a callback is processed.
The side-effect is that it prevents that `asyncshell.deliver` is called a second time (references no longer exist), but I couldn't find any case where this happened.

I would also move the `mkdir -p` call into `asyncshell.request` to recover automatically if the folder gets removed (human action or otherwise).
Another possibility would be to simply drop the folder and use `/tmp/asyncshell_reqXXXXX`. This way you don't need to worry about recovering. Even if a file is removed, in the short time window it exists, it will only affect that call to `asyncshell.request` and not future calls.

**Comment by worron** [comment]
By the way, If you are interesting in `asyncshell` util which works without saving data on disk, take a look on [my modification](https://github.com/worron/redflat/blob/master/asyncshell.lua?ts=4), it may be useful.

**Comment by copycat-killer** [comment]
@worron Your fork implements the PID table system like Linux, right?

Do you mind doing a pull a request? It would be a first solution.

**Comment by worron** [comment]
Em, only thing I did - pushing shell command output back to awesome without using temporary files.  Original PID table system wasn't changed.

I removed `asyncshell.demand` function since it never used it in my config (and didn't see it in Lain module). It seems like "dead weight" for me. Also it is possible to set time limit to request now, but this feature is fully optional and safe to remove.

Anyway, I will do a pullrequest. Take a look when you will have free time, then decide if this solution is acceptable.
